src/patterns/kinesis-lambda.ts

```diff
@@ -0,0 +1,135 @@
+ *
+ * **Example Usage**
+ *
+ * When migrating a CloudFormation stack which includes the following resource:
+ * ```yaml
+ * MyCloudFormedKinesisStream:
+ *   Type: AWS::Kinesis::Stream
+ * ```
+ * Inherit the Kinesis stream (rather than creating a new one) using:
+ * ```typescript
+ * existingKinesisStream: { logicalIdFromCloudFormation: "MyCloudFormedKinesisStream" }
+ * ```
+ *
+ * Alternatively, reference a Kinesis stream which belongs to another stack or pattern using:
+ * ```typescript
+ * existingKinesisStream: { externalKinesisStreamName: "KinesisStreamFromAnotherStack" }
+ * ```
+ */
+export interface ExistingKinesisStream {
+  logicalIdFromCloudFormation?: string;
+  externalKinesisStreamName?: string;
+}
+
+/**
+ * Configuration options for the [[`GuKinesisLambda`]] pattern.
+ *
+ * For all lambda function configuration options, see [[`GuFunctionProps`]].
+ *
+ * The `existingKinesisStream` property can be used to inherit or reference a Kinesis stream which
+ * has been created outside of this pattern (i.e. via CloudFormation, or via a different `cdk` pattern, or stack).
+ * For more details and example usage, see [[`ExistingKinesisStream`]].
+ * If this property is omitted, the [[`GuKinesisLambda`]] pattern will create a new stream.
+ *
+ * If you have specific stream configuration requirements (e.g. data retention period), these can be set via
+ * `kinesisStreamProps`.
+ *
+ * If you need to override the default stream processing options (e.g. batch size and parallelization), pass
+ * [[`StreamProcessingProps`]] via `processingProps`.
+ *
+ * You must provide `errorHandlingConfiguration` to this pattern. To prevent the lambda from repeatedly
+ * retrying to process the same records (effectively stalling stream processing), you should configure
+ * retry conditions via [[`StreamErrorHandlingProps`]]. If your use-case dictates that you must retry processing until
+ * a record expires, see [[`BlockProcessingAndRetryIndefinitely`]] for more details.
+ *
+ * It is advisable to configure an alarm based on the lambda's error percentage.
+ * To do this, add the `monitoringConfiguration` property. The required properties for this are:
+ *
+ * ```typescript
+ * monitoringConfiguration: {
+ *   toleratedErrorPercentage: <sensible_error_percentage_threshold>,
+ *   snsTopicName: "my-topic-for-cloudwatch-alerts",
+ * }
+ * ```
+ * Other alarm properties (e.g. alarm name and description) will be pre-populated with sensible defaults.
+ * For a full list of optional properties, see [[`GuLambdaErrorPercentageMonitoringProps`]].
+ *
+ * If your team do not use CloudWatch, it's possible to opt-out with the following configuration:
+ * ```typescript
+ *  monitoringConfiguration: { noMonitoring: true } as NoMonitoring
+ * ```
+ */
+export interface GuKinesisLambdaProps extends Omit<GuFunctionProps, "rules" | "apis" | "errorPercentageMonitoring"> {
+  monitoringConfiguration: NoMonitoring | GuLambdaErrorPercentageMonitoringProps;
```

Comment:
We could consider different options for alarms here (e.g. alarms based on Kinesis metrics, or on the number of dead letters if a queue is configured), but as this PR is already pretty big we might want to tackle that as a second card/PR?